### ISSUE TITLE
[Question] Bindings & TypeScript Transformers

### ISSUE THREAD

**Issue by LukasMachetanz** [nonara/ts-patch#25]
Hey!

Let's assume I am adding an import with a transformer like `import { whatever } from "whatever";`.
Somewhere else this import should be used like `whatever()`.

Another example would be adding a function like `function myFunction()` and using it somewhere else like `myFunction()`.

I fear this scenarios are not working because the actual bindings are not in place. Is this right?

I think I found a corresponding stack overflow question: https://stackoverflow.com/q/49139601/13513418
I guess it handles the same problem, right?

I like the provided answer: https://stackoverflow.com/a/49165784/13513418

But I am actually wondering how this would work with ts-patch. Is https://www.npmjs.com/package/ts-patch#transforming-program the approach this could be achieved with?

So basically I have two questions:
- How are bindings working within typescript transformers and are my examples just not possible to achieve with them?
- Is it somehow possible to achieve what I explained with ts-patch?

Thanks in advance.

**Comment by nonara** [comment]
Hi Lukas!

The issue the person is running into in that stackoverflow question is because he is trying to directly refer to something that has been imported, but because his output compilation module setting is commonjs, what he's referring to will be renamed, which causes his call not to work.

After his transformer is run, the compiler's internal transformers go through some additional steps for CommonJS which mangles the names. As Titian explained in his answer to that question, the compiler goes through steps where it walks the nodes and constructs its types and symbols first. The transform logic is performed after that, and it notably _does not_ automatically update or provide a means for it to update the types or symbols. For that reason, symbol information is unavailable for newly inserted or modified nodes. 

That has been a prevalent issue and is what lead to my adding the 'Program Transformer' functionality. You were on target with your assumptions!

I'm not sure what your exact use case is, but you should not run into the problem if—for example—you inserted a function declaration and then referred to that function in the _same file_. If, however, you wanted to also insert an import declaration and a call to it in _another_ module (and if your output format is CommonJS), then you would encounter problems.

If you do have that problem, there are two possible solutions. 

First, you could use a program transformer, performing the transformation and re-creating the program afterward. This would work, as recreating it causes the types to be walked again. The Program instance that `ts-patch` hands off to the compiler is the ultimately the one that you return, which will be the new instance with all of the correct "bindings".

The second option, which is probably better, is to simply make your program an "after" transformer. This gives your transformer the JS source code _after_ it's been converted to CommonJS, which allows you to find the proper identifier for the final import name and insert a call reference directly to it. You can still access the original TypeScript code by walking up the `original` property chain on the nodes.

Either will work and the Program transformer may be a bit easier, but in general I advise only using it if you absolutely need to, the reason being that recreating the program instance adds some extra time to the build process, and it's generally best to never add any additional drag if there's a better way.

Hope that helps!

**Comment by nonara** [comment]
FYI - I added the discussions panel to this repository, so feel free to ask future questions there.

**Comment by LukasMachetanz** [comment]
Thanks for the response. I create a discussion in the panel: https://github.com/nonara/ts-patch/discussions/27#discussion-2019093. Maybe you find the time to help me again. If there is something I could do in advance let me know. I really really appreciate your help. Thank you.
